### PR TITLE
enable empty batch for all flavors of convolution

### PR DESCRIPTION
**Pull request by ngimel**
Enables convolution with an empty batch or number of channels for all flavors of convolution (grouped convolution, convTranspose). Would make #31658 unnecessary. 
There's a philosophical question of whether we need to return zero gradients for weights if weights did not materially participate in the computation, or is it ok to leave weight gradients as None. I vote for the latter, the former is not easy to achieve without modifying code in all the backends, which is what #31658 does, and #26214 did before it, but it also does not add much value. If we decide that returning zero gradients for weights is important, I still think it should be done in the backend-independent manner, by having backend-independent empty_conv_forward function with its defined backward. A strike against #26214 approach is that it enables (meaningless) double backward through weights, ok, but any further backwards will break because returned gradients for double-backward are `zeros_like`, disconnected from the graph, so composability ends at double backward.
I'm returning a view of empty input, I'm open to the idea that it should be a clone, I also don't see how it matters if the input is empty anyway ;-)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit d8bc2bcb:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=d8bc2bcb8656c08604edafe5cb102b37256ff8c7).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4340126) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90693478) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90693478))
<details>
<summary>
<code>Jan 24 22:04:00 RuntimeError: test_autograd failed!</code>
</summary>

```
Jan 24 22:03:59 Ran 1047 tests in 264.011s 
Jan 24 22:03:59  
Jan 24 22:03:59 FAILED (errors=1, skipped=13, expected failures=1) 
Jan 24 22:03:59  
Jan 24 22:03:59 Generating XML reports... 
Jan 24 22:04:00 Traceback (most recent call last): 
Jan 24 22:04:00   File "test/run_test.py", line 456, in <module> 
Jan 24 22:04:00     main() 
Jan 24 22:04:00   File "test/run_test.py", line 449, in main 
Jan 24 22:04:00     raise RuntimeError(message) 
Jan 24 22:04:00 RuntimeError: test_autograd failed! 
Jan 24 22:04:00 + cleanup 
Jan 24 22:04:00 + retcode=1 
Jan 24 22:04:00 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by ngimel** [comment]
Test failures are real, but are related to not returning weight gradients, which, I maintain, should not be returned in this case.

**Comment by ezyang** [comment]
> I'm returning a view of empty input, I'm open to the idea that it should be a clone, I also don't see how it matters if the input is empty anyway ;-)

I think it should be a clone, so that a subsequent `resize_` doesn't affect the original tensor.

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -577,10 +565,17 @@ at::Tensor _convolution(
...
   // mkldnn conv2d weights could have been re-ordered to 5d by
   // mkldnn_reorder_conv2d_weight
+  c10::IntArrayRef weight_sizes = weight.sizes();
   if (input_is_mkldnn && (k == input.ndimension() + 1)) {
     k = input.ndimension();
+    std::vector<int64_t> weight_sizes_mkl(k);
+    weight_sizes_mkl[0] = weight.size(0) * weight.size(1);
+    std::copy_n(
+        weight.sizes().cbegin() + 2, k - 1, weight_sizes_mkl.begin() + 1);
+    weight_sizes = c10::IntArrayRef(weight_sizes_mkl);
```
Comment (ezyang) [inline aten/src/ATen/native/Convolution.cpp]: **WOOP WOOP OWNERSHIP POLICE**. IntArrayRef is a NON-OWNING reference to the pointed to array. You only allocated `weight_sizes_mkl` inside this conditional, which means it goes dead when you exit!

**Review by ezyang — CHANGES_REQUESTED** [review]
I don't think viewing the input tensor is appropriate.  If I subsequently `resize_` on the empty tensor, if I have a view on the input tensor, the resize will now expand to cover the old memory from the input tensor. That is a very observable difference!

File: aten/src/ATen/native/Convolution.cpp (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/Convolution.cpp]: oops, thanks!

**Comment by ngimel** [comment]
I've fixed scope for weight_sizes_mkl and am returning a clone. What do you think about weight gradients? I have a world ugliest workaround to automatically return infinitely differentiable zero gradients for parameters, but I'd prefer not to use it unless we decide it's necessary. @ppwwyyxx seems to want zero gradients [here](https://github.com/pytorch/pytorch/issues/12013#issuecomment-569876108)

**Comment by ppwwyyxx** [comment]
My main motivation for using zero gradients is to make DDP happy. DDP currently cannot sync gradients properly if `p.grad` is None in some but not all workers. However, alternatives such as `horovod.DistributedOptimizer` seems to handle such cases well. As a result we use hacks such as https://github.com/facebookresearch/detectron2/blob/2292cb3f24cea3d5f77f36dbd8aaf1b8d8d14e08/detectron2/layers/wrappers.py#L101-L104 to please DDP.

If the above issue is solved, then I'm not against None gradients. The only issue I see with it is inconsistency, or a lack of precise semantics -- since it's not clear when None should be used and when zeros should be used. For example, currently `(x+y).sum().backward()` where `x.shape=[0,3], y.shape=[1,3]` will produce zero gradients for y.

**Review by albanD — COMMENTED** [review]
>  I'm returning a view of empty input, I'm open to the idea that it should be a clone, I also don't see how it matters if the input is empty anyway ;-)

Autograd inplace logic 🎉 
But given that you discard the base of the view, it should not be a problem.


Given that this is a linear operation. Could a formula like this be simpler (given that you do reduction on the weights anyway):
```
out = (inp.sum() * w.sum() + b.sum()).view(1, 1, 1, 1).expand(output_size)
```

**Comment by ngimel** [comment]
Yeah I was thinking about simplifying the formula, and that should work (especially given that otherwise I'd have to add special casing for group convolutions that don't work yet. Thanks!

**Comment by ngimel** [comment]
Closing, a merge gone wrong.

**Pull request closed without merging**